Source/RxComponents/LMCCavityModes.cc

```diff
@@ -61,15 +61,31 @@ namespace locust
             }
         }
 
+
+        fVoltagePhase.resize(2); // max nchannels.  TO-DO:  Configure this as LMCGenerator::fNChannels.
+        for (int i = 0; i < GetNCavityModes(); i++)
+        {
+            fVoltagePhase[i].resize(GetNCavityModes());
+            for (int j = 0; j < GetNCavityModes(); j++)
+            {
+            	fVoltagePhase[i][j].resize(GetNCavityModes());
+            	for (int k = 0; k<GetNCavityModes(); k++)
+            	{
+            		fVoltagePhase[i][j][k].resize(GetNCavityModes());
+            	}
+            }
+        }
+
+
     	return true;
     }
 
-	bool CavityModes::AddOneModeToCavityProbe(Signal* aSignal, std::vector<double> particleXP, double excitationAmplitude, double EFieldAtProbe, std::vector<double> cavityDopplerFrequency, double dt, double phi_LO, double totalScalingFactor, unsigned sampleIndex, int channelIndex, bool initParticle)
+	bool CavityModes::AddOneModeToCavityProbe(int l, int m, int n, Signal* aSignal, std::vector<double> particleXP, double excitationAmplitude, double EFieldAtProbe, std::vector<double> cavityDopplerFrequency, double dt, double phi_LO, double totalScalingFactor, unsigned sampleIndex, int channelIndex, bool initParticle)
 	{
 		double dopplerFrequency = cavityDopplerFrequency[0];  // Only one shift, unlike in waveguide.
-		SetVoltagePhase( GetVoltagePhase(channelIndex) + dopplerFrequency * dt, channelIndex ) ;
+		SetVoltagePhase( GetVoltagePhase(channelIndex, l, m, n)[channelIndex][l][m][n] + dopplerFrequency * dt, channelIndex, l, m, n ) ;
 		double voltageValue = excitationAmplitude * EFieldAtProbe;
-		voltageValue *= cos(GetVoltagePhase(channelIndex));
+		voltageValue *= cos(GetVoltagePhase(channelIndex, l, m, n)[channelIndex][l][m][n]);
```

Comment:
You've defined GetVoltagePhase as a 4D vector, so you don't need the parameters in the function call (or alternatively you could actually use the parameters in the function and return a scalar double and then not use the indices here).